### DISCUSSION TITLE
2.8.0rc1

### DISCUSSION THREAD

**Discussion by Zuzu-Typ** [Zuzu-Typ/PyGLM#263] [General]
# Prerelease of the new package structure.
**Important**: PyGLM is now packaged as
+ PyGLM (root package) 
  + .glm (library) 
  + .glm_typing (typing stubs)
+ glm (for backwards compatibility)
  
Please test this new package structure with your projects and give me feedback in the discussion for this release.

Update your import to this for compatibility with old *and* future versions:
```Python
try:
    from PyGLM import glm
except ImportError:
    import glm
```

## What's Changed
* Make array generic. by @esoma in https://github.com/Zuzu-Typ/PyGLM/pull/172
* Updated glm submodule to latest master by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/250
* Fix compiler warnings by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/251
* Add quat_to_vec4 (and vice-versa) documentation by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/252
* Updated PyGLM packaging structure by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/203
* Added visual studio project by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/254
* Adjusted project structure by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/255
* Updated Readme and manifest by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/256
* Fixed readme by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/257
* Test on push only by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/258
* Slightly improved readme by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/259
* Bump Version (2.8.0rc1) by @Zuzu-Typ in https://github.com/Zuzu-Typ/PyGLM/pull/262

## New Contributors
* @esoma made their first contribution in https://github.com/Zuzu-Typ/PyGLM/pull/172

**Full Changelog**: https://github.com/Zuzu-Typ/PyGLM/compare/2.7.3...2.8.0rc1

<hr /><em>This discussion was created from the release <a href='https://github.com/Zuzu-Typ/PyGLM/releases/tag/2.8.0rc1'>2.8.0rc1</a>.</em>

**Comment by Zuzu-Typ** [comment]
Given that [PEP 8](https://peps.python.org/pep-0008/#package-and-module-names) states that module and package names should be lowercase I should consider renaming the root package to `pyglm`. That would comply with it.

**Reply by jimy-byerley** [reply]
By the way, why renaming the module `pyglm` instead of `glm` ?
Also why are the two submodules prefixed `glm*` ?
I think `numpy` is using the following submodule convention (with `numpy` instead of `glm` of course):
- `glm.core` with the glm functions
- `glm.typing` with the type hints

**Reply by Zuzu-Typ** [reply]
> By the way, why renaming the module `pyglm` instead of `glm` ?

Because of the module [glm](https://pypi.org/project/glm/). It's been causing confusion since the beginning of PyGLM that the import is called `glm`, but the module is called `PyGLM` (see https://github.com/Zuzu-Typ/PyGLM/issues/25). This also means that you could not install [glm](https://pypi.org/project/glm/) if you also want [PyGLM](https://pypi.org/project/PyGLM/) installed. 
I think it's a flaw of pypi that they allow module names and package names to be different in the first place.

> Also why are the two submodules prefixed `glm*` ?

The submodule glm is a single compiled file (`glm.pyd`) containing the code for PyGLM.
The submodule glm_typing (`glm_typing.py`) contains type aliases for writing type annotations. This is from [pyglm-typing](https://github.com/esoma/pyglm-typing/blob/main/src/glm_typing/__init__.py). It's simply how the submodule of pyglm-typing is called.
I also added an alias for glm_typing which is simply called `typing`, so you could also see the structure as:
+ `pyglm`
  + `.glm`
  + `.typing`

**Reply by jimy-byerley** [reply]
So [glm](https://pypi.org/project/glm/) is a real thing ... always tough it was some kind of crappy forgotten module ^.^'

> I think it's a flaw of pypi that they allow module names and package names to be different in the first place.

definitely, yes. They should have done it juste like [crates.io](https://crates.io) where you put your source into a folder `src` which is used to generate the library with the name provided in the manifest ... or at least check that the folder name is matching the name in the manifest

**Reply by Zuzu-Typ** [reply]
Alright, I've now renamed the module to `pyglm`
